File: src/core/execution/nodes/memory_selector.py
"""
Memory Selector Node
Selects relevant conversation context from storage
"""
from typing import Dict, Any, List, Optional
from pathlib import Path
import json
from ..node_base import BaseNode, ExecutionContext
from ....utils.logger import get_logger

logger = get_logger(__name__)

# LangChain is REQUIRED
try:
    from langchain_core.messages import HumanMessage, AIMessage
except ImportError as e:
    raise ImportError(
        "LangChain is required for memory management. Please install it with: pip install langchain langchain-core"
    ) from e


class MemorySelectorNode(BaseNode):
    """
    Selects relevant conversation context from storage
    
    Handles:
    - Loading recent interactions from storage
    - Loading summaries from storage
    - Using MemorySelector agent to intelligently select relevant memories
    - Formatting output as ConversationContextDict
    
    Inputs:
        query: User query string (for context selection)
        storage_instance: StorageInterface instance (from Memory Storage Node)
        user_id: User identifier (optional)
        llm: ObeliskLLM instance (optional, defaults to container's LLM)
        enable_recent_buffer: Whether to include recent conversation buffer (optional, default: True)
        k: Number of recent message pairs to keep in buffer (optional, default: 10)
    
    Outputs:
        context: ConversationContextDict with 'messages' and 'memories'
    """
    
    # Class-level cache for buffer managers per storage instance
    _buffer_managers: Dict[str, Any] = {}
    
    def __init__(self, node_id: str, node_data: Dict[str, Any]):
        """Initialize memory selector node"""
        super().__init__(node_id, node_data)
    
    def _get_buffer_manager(self, storage_instance, k: int):
        """Get or create buffer manager for storage instance"""
        # Use storage instance id or path as cache key
        storage_id = id(storage_instance)
        if storage_id not in self._buffer_managers:
            from ....memory.buffer_manager import RecentBufferManager
            self._buffer_managers[storage_id] = RecentBufferManager(k=k)
        return self._buffer_managers[storage_id]
    
    def _load_all_summaries_from_storage(self, storage_instance, user_id: str, limit: int = 30) -> List[Dict[str, Any]]:
        """Load all summaries from storage (up to limit, most recent first)"""
        try:
            summaries_list = []
            
            # For LocalJSONStorage, summaries are in memory/activities.json
            if hasattr(storage_instance, 'memory_path'):
                activities_file = Path(storage_instance.memory_path) / "activities.json"
            elif hasattr(storage_instance, 'base_path'):
                # Fallback for old structure (backward compatibility)
                activities_file = Path(storage_instance.base_path) / "memory" / "activities.json"
                if not activities_file.exists():
                    # Try old location
                    activities_file = Path(storage_instance.base_path) / "activities.json"
            else:
                return []
            
            if activities_file.exists():
                with open(activities_file, 'r') as f:
                    activities = json.load(f)
                
                # Find all conversation_summary activities for this user
                user_summaries = [
                    activity for activity in activities
                    if activity.get('type') == 'conversation_summary'
                    and activity.get('message', '').endswith(f'user {user_id}')
                ]
                
                # Sort by created_at (most recent first) and limit
                user_summaries.sort(key=lambda x: x.get('created_at', ''), reverse=True)
                user_summaries = user_summaries[:limit]
                
                # Extract summary_data from each
                for activity in user_summaries:
                    metadata = activity.get('metadata', {})
                    summary_data = metadata.get('summary_data')
                    if summary_data:
                        # Add activity ID for reference
                        summary_data['_activity_id'] = activity.get('id')
                        summary_data['_created_at'] = activity.get('created_at')
                        summaries_list.append(summary_data)
            
            return summaries_list
            
        except Exception as e:
            logger.error(f"Error loading summaries from storage: {e}")
            return []
    
    def _select_relevant_memories(
        self,
        selector_agent,
        user_query: str,
        summaries: List[Dict[str, Any]],
        top_k: int = 5
    ) -> List[Dict[str, Any]]:
        """
        Use Memory Selector agent to select relevant memories from a list of summaries
        
        Args:
            selector_agent: MemorySelectorAgent instance
            user_query: Current user query (required)
            summaries: List of summary dictionaries
            top_k: Number of relevant memories to select (default: 5)
            
        Returns:
            List of selected relevant summary dictionaries
        """
        if not summaries:
            return []
        
        return selector_agent.select(user_query, summaries, top_k)
    
    def execute(self, context: ExecutionContext) -> Dict[str, Any]:
        """Execute memory selector node"""
        query = self.get_input_value('query', context, '')
        storage_instance = self.get_input_value('storage_instance', context, None)
        user_id = self.get_input_value('user_id', context, None)
        llm = self.get_input_value('llm', context, None)
        enable_recent_buffer = self.get_input_value('enable_recent_buffer', context, True)
        k = self.get_input_value('k', context, 10)
        
        # Resolve template variables
        if isinstance(query, str) and query.startswith('{{') and query.endswith('}}'):
            var_name = query[2:-2].strip()
            query = context.variables.get(var_name, '')
        
        if isinstance(user_id, str) and user_id.startswith('{{') and user_id.endswith('}}'):
            var_name = user_id[2:-2].strip()
            user_id = context.variables.get(var_name, None)
        
        # Convert enable_recent_buffer to boolean if it's a string
        if isinstance(enable_recent_buffer, str):
            enable_recent_buffer = enable_recent_buffer.lower() in ('true', '1', 'yes', 'on')
        enable_recent_buffer = bool(enable_recent_buffer)
        
        # Default user_id if not provided
        if user_id is None or user_id == '':
            user_id = f"user_{self.node_id}"
        
        # Default to container's LLM if not provided
        if llm is None:
            llm = context.container.llm
        
        # Validate inputs
        if not storage_instance:
            raise ValueError("storage_instance is required for MemorySelectorNode")
        
        if not query:
            raise ValueError("query is required for MemorySelectorNode")
        
        # Import agent
        from ....memory.agents.memory_selector import MemorySelector as MemorySelectorAgent
        selector_agent = MemorySelectorAgent(llm)
        
        # Get buffer manager (shared per storage instance)
        buffer_manager = self._get_buffer_manager(storage_instance, int(k))
        
        # Get conversation context
        conversation_messages = []
        memories_parts = []
        
        # Get recent messages from buffer window (recent conversation) if enabled
        # Convert to Qwen3 message format: [{"role": "user", "content": "..."}, ...]
        if enable_recent_buffer:
            buffer = buffer_manager.get_buffer(str(user_id), storage_instance)
            messages = buffer.get_messages()
            
            for msg in messages:
                if hasattr(msg, 'content'):
                    # Convert LangChain messages to Qwen3 format
                    if isinstance(msg, HumanMessage):
                        conversation_messages.append({
                            "role": "user",
                            "content": msg.content
                        })
                    elif isinstance(msg, AIMessage):
                        conversation_messages.append({
                            "role": "assistant",
                            "content": msg.content
                        })
        
        # Load all summaries and use intelligent selection (always runs)
        all_summaries = self._load_all_summaries_from_storage(storage_instance, str(user_id), limit=30)
        
        if all_summaries:
            # Use LLM to select relevant memories if we have multiple summaries
            if len(all_summaries) > 1:
                selected_summaries = self._select_relevant_memories(selector_agent, str(query), all_summaries, top_k=5)
            else:
                # Only one summary, use it as-is (zero summaries case handled by outer if)
                selected_summaries = all_summaries
            
            # Format selected memories
            for summary_data in selected_summaries:
                # Format important facts as bullet points
                important_facts = summary_data.get('importantFacts', [])
                if important_facts:
                    if not memories_parts:
                        memories_parts.append("[Memories]")
                    if isinstance(important_facts, list):
                        for fact in important_facts:
                            memories_parts.append(f"- {fact}")
                    else:
                        memories_parts.append(f"- {important_facts}")
                
                # Add user context if available (merge to avoid duplicates)
                user_context = summary_data.get('userContext', {})
                if user_context and isinstance(user_context, dict):
                    if memories_parts and "[User Context]" not in "\n".join(memories_parts):
                        memories_parts.append("")  # Add separator
                        memories_parts.append("[User Context]")
                    elif "[User Context]" not in "\n".join(memories_parts):
                        memories_parts.append("[User Context]")
                    for key, value in user_context.items():
                        # Only add if not already present
                        context_line = f"- {key}: {value}"
                        if context_line not in memories_parts:
                            memories_parts.append(context_line)
        
        memories_str = "\n".join(memories_parts) if memories_parts else ""
        
        return {
            'context': {
                "messages": conversation_messages,
                "memories": memories_str
            }
        }

File: ts/src/core/execution/nodes/memorySelector.ts
/**
 * MemorySelectorNode – loads recent conversations + relevant memories and
 * builds a conversation context dict for the InferenceNode.
 * Mirrors Python src/core/execution/nodes/memory_selector.py
 *
 * Inputs:
 *   query: User query string (required, for context selection)
 *   storage_instance: StorageInterface instance (required)
 *   user_id: User identifier (optional, defaults to user_{nodeId})
 *   model: InferenceClient instance (accepts 'model' or 'llm')
 *   enable_recent_buffer: Whether to include recent conversation buffer (default: true)
 *   k: Number of recent message pairs to keep in buffer (default: 10)
 *
 * Outputs:
 *   query: Original query (pass-through for cleaner flow)
 *   context: ConversationContextDict with 'messages' and 'memories'
 */
import { BaseNode, ExecutionContext } from "../nodeBase";
import { StorageInterface, ActivityLog } from "../../types";
import { InferenceClient, resolveInferenceClient } from "./inference/inferenceClient";
import { RecentBufferManager } from "./memory/bufferManager";
import { extractJsonFromLlmResponse } from "../../../utils/jsonParser";
import { getLogger } from "../../../utils/logger";

const logger = getLogger("memorySelector");

// Class-level cache for buffer managers, keyed by (storageInstanceId, k)
let bufferManagerIdCounter = 0;
const storageIdMap = new WeakMap<object, number>();
const bufferManagerCache: Record<string, RecentBufferManager> = {};

function getBufferManager(storage: StorageInterface, k: number): RecentBufferManager {
  let sid = storageIdMap.get(storage);
  if (sid === undefined) {
    sid = bufferManagerIdCounter++;
    storageIdMap.set(storage, sid);
  }
  const cacheKey = `${sid}_${k}`;
  if (!bufferManagerCache[cacheKey]) {
    bufferManagerCache[cacheKey] = new RecentBufferManager(k);
  }
  return bufferManagerCache[cacheKey];
}

export class MemorySelectorNode extends BaseNode {
  async execute(context: ExecutionContext): Promise<Record<string, unknown>> {
    const query = this.getInputValue("query", context, "") as string;
    const storage = this.getInputValue(
      "storage_instance",
      context,
      undefined
    ) as StorageInterface | undefined;
    let userId = this.getInputValue("user_id", context, null) as
      | string
      | null;
    // Accept both 'model' (from InferenceConfigNode) and 'llm' (legacy); unwrap { model, agent_id } shape
    const modelRaw =
      this.getInputValue("model", context, undefined) ||
      this.getInputValue("llm", context, undefined);
    const agentId =
      modelRaw &&
      typeof modelRaw === "object" &&
      "agent_id" in modelRaw
        ? (modelRaw as { agent_id?: string }).agent_id
        : undefined;
    const model = resolveInferenceClient(modelRaw);
    let enableRecentBuffer = this.getInputValue(
      "enable_recent_buffer",
      context,
      true
    );
    const kRaw = Number(this.getInputValue("k", context, 10));
    const k = Number.isFinite(kRaw) ? kRaw : 10;

    // Resolve template variables
    const queryStr = String(query ?? "");

    // Convert enable_recent_buffer to boolean
    if (typeof enableRecentBuffer === "string") {
      enableRecentBuffer = ["true", "1", "yes", "on"].includes(
        (enableRecentBuffer as string).toLowerCase()
      );
    }
    enableRecentBuffer = Boolean(enableRecentBuffer);

    // Default user_id if not provided
    if (!userId || userId === "") {
      userId = `user_${this.nodeId}`;
    }

    // Validate inputs
    if (!storage) {
      throw new Error(
        "storage_instance is required for MemorySelectorNode"
      );
    }
    if (!queryStr) {
      throw new Error("query is required for MemorySelectorNode");
    }
    if (!model) {
      throw new Error(
        "model/llm is required for MemorySelectorNode. Connect an InferenceConfigNode."
      );
    }

    // Validate k (ensure no NaN propagation)
    const kInt = Math.max(1, Math.floor(k));

    // ── Recent conversation buffer ──────────────────────────────────

    const conversationMessages: Array<{ role: string; content: string }> = [];

    if (enableRecentBuffer) {
      const bufMgr = getBufferManager(storage, kInt);
      const buffer = await bufMgr.getBuffer(String(userId), storage);
      const messages = buffer.getMessages();

      logger.debug(
        `[MemorySelector] Buffer enabled: loaded ${messages.length} messages for user_id=${userId}`
      );

      for (const msg of messages) {
        if (msg.role === "human") {
          conversationMessages.push({ role: "user", content: msg.content });
        } else if (msg.role === "ai") {
          conversationMessages.push({
            role: "assistant",
            content: msg.content,
          });
        }
      }
    } else {
      logger.debug(
        `[MemorySelector] Buffer disabled for user_id=${userId}`
      );
    }

    // ── Load summaries and select relevant memories ─────────────────

    const memoriesParts: string[] = [];

    const allSummaries = await this._loadAllSummaries(
      storage,
      String(userId),
      30
    );

    logger.debug(
      `[MemorySelector] Loaded ${allSummaries.length} summaries for user_id=${userId}`
    );

    if (allSummaries.length > 0) {
      let selectedSummaries: Array<Record<string, unknown>>;

      if (allSummaries.length > 1) {
        selectedSummaries = await this._selectRelevantMemories(
          model,
          agentId,
          queryStr,
          allSummaries,
          5
        );
      } else {
        selectedSummaries = allSummaries;
      }

      // Format selected memories (mirrors Python formatting)
      for (const summaryData of selectedSummaries) {
        const importantFacts =
          (summaryData.importantFacts as unknown[]) ?? [];
        if (importantFacts.length) {
          if (!memoriesParts.length) memoriesParts.push("[Memories]");
          for (const fact of importantFacts) {
            if (typeof fact === "object" && fact !== null) {
              const vals = Object.values(fact as Record<string, unknown>);
              memoriesParts.push(
                `- ${vals.length ? String(vals[0]) : String(fact)}`
              );
            } else {
              memoriesParts.push(`- ${String(fact)}`);
            }
          }
        }

        const userContext =
          (summaryData.userContext as Record<string, unknown>) ?? null;
        if (userContext && typeof userContext === "object") {
          const joined = memoriesParts.join("\n");
          if (!joined.includes("[User Context]")) {
            if (memoriesParts.length) memoriesParts.push(""); // separator
            memoriesParts.push("[User Context]");
          }
          for (const [key, value] of Object.entries(userContext)) {
            const line = `- ${key}: ${value}`;
            if (!memoriesParts.includes(line)) {
              memoriesParts.push(line);
            }
          }
        }
      }
    }

    const memoriesStr = memoriesParts.length
      ? memoriesParts.join("\n")
      : "";

    const contextOutput = {
      messages: conversationMessages,
      memories: memoriesStr,
    };

    logger.debug(
      `[MemorySelector] Final context for user_id=${userId}: ${conversationMessages.length} messages, ${memoriesStr.length} chars of memories`
    );

    return {
      query: queryStr, // Pass through original query
      context: contextOutput,
    };
  }

  // ── Private helpers ──────────────────────────────────────────────

  private async _loadAllSummaries(
    storage: StorageInterface,
    userId: string,
    limit: number
  ): Promise<Array<Record<string, unknown>>> {
    try {
      const activities: ActivityLog[] = await storage.getActivityLogs(
        "conversation_summary",
        limit * 2
      );

      const userSummaries: Array<Record<string, unknown>> = [];

      for (const activity of activities) {
        const meta = (activity.metadata ?? {}) as Record<string, unknown>;
        const summaryData = (meta.summary_data ?? {}) as Record<
          string,
          unknown
        >;
        const activityUserId =
          summaryData.user_id ?? meta.user_id ?? null;

        // Match user_id if found in metadata, otherwise fallback to message pattern
        let matches = false;
        if (activityUserId && String(activityUserId) === userId) {
          matches = true;
        } else if (!activityUserId) {
          // Fallback: check message pattern
          if ((activity.message ?? "").endsWith(`user ${userId}`)) {
            matches = true;
          }
        }

        if (matches && summaryData && Object.keys(summaryData).length > 0) {
          userSummaries.push({
            ...summaryData,
            _activity_id: activity.id,
            _created_at: activity.created_at,
          });
          if (userSummaries.length >= limit) break;
        }
      }

      return userSummaries;
    } catch (err) {
      logger.error(`[MemorySelector] Error loading summaries: ${err}`);
      return [];
    }
  }

  private async _selectRelevantMemories(
    model: InferenceClient,
    agentId: string | undefined,
    userQuery: string,
    summaries: Array<Record<string, unknown>>,
    topK: number
  ): Promise<Array<Record<string, unknown>>> {
    if (!summaries.length) return [];
    if (summaries.length <= topK) return summaries;

    try {
      // Format summaries (mirrors Python formatting)
      let summariesText = "";
      for (let i = 0; i < summaries.length; i++) {
        const summary = summaries[i];
        let str = `Memory ${i}:\n`;
        str += `  Summary: ${summary.summary ?? "N/A"}\n`;

        const topics = (summary.keyTopics as unknown[]) ?? [];
        const topicStrs = topics.map((t) =>
          typeof t === "object" && t !== null
            ? String(
                Object.values(t as Record<string, unknown>)[0] ??
                  String(t)
              )
            : String(t)
        );
        str += `  Topics: ${topicStrs.join(", ")}\n`;

        const facts = (summary.importantFacts as unknown[]) ?? [];
        const factStrs = facts.map((f) =>
          typeof f === "object" && f !== null
            ? String(
                Object.values(f as Record<string, unknown>)[0] ??
                  String(f)
              )
            : String(f)
        );
        str += `  Facts: ${factStrs.join(", ")}\n`;

        const userCtx =
          (summary.userContext as Record<string, unknown>) ?? {};
        if (userCtx && Object.keys(userCtx).length) {
          str += `  Context: ${Object.entries(userCtx)
            .map(([k, v]) => `${k}=${v}`)
            .join(", ")}\n`;
        }
        summariesText += str + "\n";
      }

      const systemPrompt = `You are a memory selector. Your role is to analyze memories and select the ${topK} most relevant ones for a user query.

You MUST return ONLY valid JSON. No markdown code blocks, no explanations, no text before or after the JSON. Start with { and end with }.

Analyze which memories are most relevant to the user query and return a JSON object with:
- selected_indices: Array of 0-based indices of the ${topK} most relevant memories (e.g., [0, 2, 5])
- reason: Brief explanation of why these memories were selected

Example of correct JSON format:
{
  "selected_indices": [0, 2, 5],
  "reason": "Memory 0 discusses the main topic, Memory 2 contains relevant context, Memory 5 has related facts"
}`;

      const query = `User Query: ${userQuery}\n\nAvailable Memories:\n${summariesText}\n\nReturn the indices (0-based) of the ${topK} most relevant memories. Return ONLY the JSON object, nothing else.`;

      const result = await model.generate(
        query,
        systemPrompt,
        0.1, // Low quantum_influence for consistent selection
        800,
        null,
        false,
        agentId
      );

      const selectionText = (result.response ?? "").trim();
      const raw = extractJsonFromLlmResponse(
        selectionText,
        "memory selection"
      );
      const selectionData = Array.isArray(raw) ? ({} as Record<string, unknown>) : raw;

      const selectedIndices =
        (selectionData.selected_indices as number[]) ?? [];

      const selectedMemories: Array<Record<string, unknown>> = [];
      for (const idx of selectedIndices) {
        if (typeof idx === "number" && idx >= 0 && idx < summaries.length) {
          selectedMemories.push(summaries[idx]);
        }
      }

      if (selectedMemories.length) {
        logger.debug(
          `[MemorySelector] Selected ${selectedMemories.length} relevant memories from ${summaries.length} total`
        );
        return selectedMemories;
      }

      // Fallback: return first topK if selection failed
      logger.warning(
        `[MemorySelector] Memory selection returned invalid indices: ${JSON.stringify(selectedIndices)}, falling back to first ${topK}`
      );
      return summaries.slice(0, topK);
    } catch (err) {
      logger.error(`[MemorySelector] Error in memory selection: ${err}`);
      // Fallback: return first topK
      return summaries.slice(0, topK);
    }
  }
}
